[Rhosinfra-23] dynamically create default.ini from spec file

obaranov opened the pull request:
1) Now default values can be defined in the spec files
2) Removed default.ini file
3) Added generate-conf-file option to generate ini file from the defaults values. 

Several comments: 
 - We cannot remove 'from-file' from spec file because clg will raise then the unsupported argument exception. Anyway from-file argument does not affect default values logic anymore.  
 - dafault.ini file is no longer required. All default values will be stored in spec files. These default values can be overriden by env variables and by the ini file. 

@yfried-redhat please review.

yfried-redhat commented on cli/conf.py:
```diff
@@ -1,5 +1,7 @@
+import sys
+
 import clg
 import os
```
AFAIK, imports should be grouped:
1. builtin
2. 3rd party
3. local imports

yfried-redhat commented on cli/conf.py:
```diff
@@ -74,17 +74,35 @@ def parse_args(cls, module_name, config, args=None):
+            except Exception as ex:
+                raise exceptions.IRException(ex.message)
+            finally:
+                sys.exit(0)
```
I don't think we should have sys.exit anywhere here.
I suggest setting `dry-run=True` and replacing output file with the rest of the the output.
Alternatively, generate-conf-file can be a new subcommand, with "virsh" as its input.

yfried-redhat replied:
User should be unable to use generate-conf-file with any other virsh argument (as it has no meaning). Except, maybe, for `-o`/`--output`

obaranov replied:
@yfried-redhat can we move that comment to the separate task? This requires major logic change. I'd prefer to implement that as a separate feature.

yfried-redhat commented on cli/conf.py:
```diff
@@ -105,6 +123,81 @@ def parse_args(cls, module_name, config, args=None):
         return res_args
 
+
+class SpecLoader(object):
```
Do we need a whole class of "class methods"? Seems like it would be simpler to make this class a different module (spec.py) with simple functions.

yfried-redhat commented on cli/conf.py:
```diff
@@ -105,6 +123,81 @@ def parse_args(cls, module_name, config, args=None):
+        return clg_options, global_options, subcommands_options
+
+    @classmethod
+    def _get_options(cls, spec_dict, lookup=False, trim=False):
```
This need some documentation.
What's the diff between `_get_options` and `__get_options`?

yfried-redhat commented on cli/conf.py:
```diff
@@ -105,6 +123,81 @@ def parse_args(cls, module_name, config, args=None):
+                cls.__get_options(options, group, lookup, trim)
+
+    @classmethod
+    def __get_options(cls, options, spec_dict, lookup, trim):
```
Need doc

yfried-redhat commented on cli/conf.py:
```diff
@@ -105,6 +123,81 @@ def parse_args(cls, module_name, config, args=None):
+                    cls.trim_option(opt_name, opt_params)
+
+    @classmethod
+    def lookup_option(cls, opt_name, opt_params):
```
ditto

yfried-redhat commented on cli/conf.py:
```diff
@@ -105,6 +123,81 @@ def parse_args(cls, module_name, config, args=None):
+                opt_params['default'])
+
+    @classmethod
+    def trim_option(cls, opt_name, opt_params):
```
ditto

yfried-redhat replied:
Guessing: "removes clg attributes (default and required) from clg argument."
If so, please rename "opt_name" and "opt_params". If not, please provide a proper docstring

yfried-redhat commented on cli/conf.py:
```diff
@@ -105,6 +123,81 @@ def parse_args(cls, module_name, config, args=None):
+        return global_options, subcommand_options
+
+    @classmethod
+    def __get_groups_options(cls, options, spec_dict, lookup, trim):
```
Need doc

yfried-redhat commented on cli/utils.py:
```diff
@@ -49,9 +46,9 @@ def none_resolver(first, second, key):
+            elif second[key] is not None:
                 first[key].append(first[key])
 
         if key not in first or first[key] is None:
```
Comments about dict_merge:

- dict merge strategy is unclear. need more doc. testing would be helpful as well.
- let's avoid merging lists. values should be overwritten
- the recursive call to dict_merge doesn't use the conflict resovler
- Unless we have a need for different conflict resolvers, I believe that we can allways use the default one.
- If I understand correctly, first[key] is chosen unless it is `None`, in which case, second[key] is chosen. I suggest to always use use second[key] unless:
  - key not in second
  - both are dicts and then invoke recursion.

aopincar commented:
Hi Oleksii,

Regards ticket #33 in Jira:
I have a unittest ready for dicts merging:  [https://github.com/aopincar/InfraRed/tree/RHOSINFRA-33](url)

About the dict merge, ticket #34 in Jira, I'll try to push a solution by the morning.

obaranov commented:
Review comments addressed: 
1) Spec related code was moved to the separate module.
2) Added method's documentation
3) Replaced conflict resolver for the dict_merger method. None-resolver is still required to merger cli, env, anc ini arguments. List merge is also required for the cli arguments like --extra-vars with action=append. 
4) Added unit tests for dict_merge (including @aopincar test)

gamado commented:
retest
